src/mds/Migrator.h

```diff
@@ -266,14 +270,42 @@ class Migrator {
   struct export_state_t {
     export_state_t() {}
 
-    int state = 0;
+    void set_state(int s) {
+      ceph_assert(s != state);
+      if (state != EXPORT_NONE) {
+	auto& t = state_history.at(state);
+	t.second = double(ceph_clock_now()) - double(t.first);
+      }
+      state = s;
+      state_history[state] = std::pair<utime_t, double>(ceph_clock_now(), 0.0);
+    }
+    utime_t get_start_time(int s) const {
+      ceph_assert(state_history.count(s) > 0);
+      return state_history.at(s).first;
+    }
+    double get_time_spent(int s) const {
+      ceph_assert(state_history.count(s) > 0);
+      const auto& t = state_history.at(s);
+      return s == state ? double(ceph_clock_now()) - double(t.first) : t.second;
+    }
+    double get_freeze_tree_time() const {
+      ceph_assert(state >= EXPORT_DISCOVERING);
+      ceph_assert(state_history.count(EXPORT_DISCOVERING) > 0);
+      return double(ceph_clock_now()) - double(state_history.at(EXPORT_DISCOVERING).first);
+    };
+
+    int state = EXPORT_NONE;
     mds_rank_t peer = MDS_RANK_NONE;
     uint64_t tid = 0;
     std::set<mds_rank_t> warning_ack_waiting;
     std::set<mds_rank_t> notify_ack_waiting;
     std::map<inodeno_t,std::map<client_t,Capability::Import> > peer_imported;
     MutationRef mut;
     size_t approx_size = 0;
+    // record the start time and time spent of each export state
+    std::map<int, std::pair<utime_t, double> > state_history;
```

Comment:
just curious to know why `state_history` can't be an `std::array<>` or a `std::vector<>` instead of a `std::map<>` ... since the states advance linearly to the next state ?

is that just to future-proof the code w.r.t. addition of new states ?

Reply:
> just curious to know why `state_history` can't be an `std::array<>` or a `std::vector<>` instead of a `std::map<>` ... since the states advance linearly to the next state ?
> 
> is that just to future-proof the code w.r.t. addition of new states ?

I think vector can also be used here. I use map because the state value can be used as the key and it's convenient to query the value.

Reply:
well, the state value can be used as an index into the vector<> assuming that EXPORT_NONE isn't represented as a state in the vector<> since its value is -1

Reply:
Do we really need `EXPORT_NONE` state. I think that can be done away with.

Reply:
> Do we really need `EXPORT_NONE` state. I think that can be done away with.

Yes, I think you are right. I have deleted `EXPORT_NONE`. Thanks!